### PR TITLE
✨ Change API for OpenStackMachine.Spec.Image

### LINKED ISSUE #1795: Convert `Image` & `ImageUUID` into a single param using an `ImageFilter` instead
/kind feature

**Describe the solution you'd like**
Right now, the image for an `OpenStackMachine` can be defined either its name via `Spec.Image` or via `Spec.ImageUUID`.
We could simply this by using a single parameter which would be of type `ImageFilter`.

### PR DESCRIPTION
**Pull request by EmilienM**
**What this PR does / why we need it**:

Right now, the image for an OpenStackMachine can be defined either
its name via Spec.Image or via Spec.ImageUUID.

Now we will use a single parameter which would be of type ImageFilter.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1795

**TODOs**:

- [X] squashed commits
- [X] includes documentation
- [X] adds unit tests

### CODE REVIEW

**Comment by EmilienM** [comment]
Something is wrong in the conversion, the bastion instanceSpec gets wrong API:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1796/pull-cluster-api-provider-openstack-e2e-full-test/1737146179717697536/artifacts/clusters/clusterctl-upgrade-5awqmp/clusterctl-upgrade-7iukbs-cluster-template.yaml

```yaml
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha6
kind: OpenStackCluster
metadata:
  name: clusterctl-upgrade-7iukbs
  namespace: clusterctl-upgrade
spec:
  allowAllInClusterTraffic: true
  apiServerLoadBalancer:
    enabled: true
  bastion:
    enabled: true
    instance:
      flavor: m1.tiny
      image:
        name: cirros-0.6.1-x86_64-disk
```

**Comment by EmilienM** [comment]
I need to move `test/e2e/data/kustomize/common-patches/cni/patch-cluster.yaml` to versioned patches so it's not applied on v1alpha8.

File: pkg/cloud/services/compute/instance.go
Comment (jichenjc) [inline pkg/cloud/services/compute/instance.go:573]: should we add the filter info into list?

**Review by jichenjc — COMMENTED** [review]
the others lgtm

File: pkg/cloud/services/compute/instance.go
Comment (jichenjc) [inline pkg/cloud/services/compute/instance.go:566]: should we follow https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1779/files#diff-2669f4e6453b138e99121b1203a91f87fa3bc88468c8873c1912c0492566075fR28 ?

File: pkg/cloud/services/compute/instance.go
Comment (EmilienM) [reply pkg/cloud/services/compute/instance.go:573]: good idea, i'm doing this.

File: pkg/cloud/services/compute/instance.go
Comment (EmilienM) [reply pkg/cloud/services/compute/instance.go:566]: good idea, i'm doing this.

**Review by dulek — APPROVED** [review]
/lgtm

**Comment by MaysaMacedo** [comment]
lgtm

File: pkg/cloud/services/compute/instance.go
Comment (stephenfin) [inline pkg/cloud/services/compute/instance.go:564]:
```suggestion
// Helper function for getting image ID from name, ID, or tags.
```

File: api/v1alpha8/types.go
Comment (stephenfin) [inline api/v1alpha8/types.go:26]:
```suggestion
	// The ID of the desired image. If this is provided, the other filters will be ignored.
	ID   string   `json:"id,omitempty"`
```

File: api/v1alpha8/types.go
Comment (stephenfin) [inline api/v1alpha8/types.go:27]:
```suggestion
	// The name of the desired image. If specified, the combination of name and tags must return a single matching image or an error will be raised.
	Name string   `json:"name,omitempty"`
```

File: api/v1alpha8/types.go
Comment (stephenfin) [inline api/v1alpha8/types.go:28]:
```suggestion
	// The tags associated with the desired image. If specified, the combination of name and tags must return a single matching image or an error will be raised.
	Tags []string `json:"tags,omitempty"`
```

File: api/v1alpha8/openstackmachine_types.go
Comment (stephenfin) [inline api/v1alpha8/openstackmachine_types.go:49]: If I understand correctly, this is no longer true: we will now use the image requested by the filter when creating the root volume? If so, how about

```suggestion
	// If the rootVolume is specified, this be used when creating the root volume.
```

**Review by stephenfin — CHANGES_REQUESTED** [review]

File: api/v1alpha8/openstackmachine_types.go
Comment (EmilienM) [reply api/v1alpha8/openstackmachine_types.go:49]: I'm not sure why we had this comment before, since the image ID was not ignored and used when creating the rootVolume:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/98f154bafb2b51318a7d3eb6655a206a491c9620/pkg/cloud/services/compute/instance.go#L451

So your suggestion is valid now is also valid without this change. I'll accept it.

File: api/v1alpha8/openstackmachine_types.go
Comment (stephenfin) [inline api/v1alpha8/openstackmachine_types.go:49]:
```suggestion
	// If the rootVolume is specified, this will be used when creating the root volume.
```

File: pkg/cloud/services/compute/instance_test.go
Comment (stephenfin) [inline pkg/cloud/services/compute/instance_test.go:149]: Do we want a test case for filtering by tags?

**Review by stephenfin — CHANGES_REQUESTED** [review]
One further nit and a question.

File: pkg/cloud/services/compute/instance_test.go
Comment (EmilienM) [reply pkg/cloud/services/compute/instance_test.go:149]: I can add, it doesn't hurt

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (stephenfin) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:27]: Sorry, this was my mistake.

```suggestion
The `ImageFilter` object allows selection of an image by name, by ID or by tags.
```

**Review by stephenfin — APPROVED** [review]

File: api/v1alpha6/conversion_test.go
Comment (EmilienM) [inline api/v1alpha6/conversion_test.go:92]: This is not correct.
@mdbooth found a problem in the conversion functions: `RestoreField` must be provided to a `HashedFieldRestorer`.

I'll wait for him to provide a fix on some other PRs and apply it here.

**Comment by EmilienM** [comment]
CI is red until we get https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1812

**Review by mdbooth — COMMENTED** [review]
Please could you add the resolved image reference to `ReferencedMachineResources` now we've got it?

File: controllers/openstackcluster_controller.go
Comment (EmilienM) [inline controllers/openstackcluster_controller.go:399]: I've put that here because if the user changes the bastion spec with a new image, the instanceSpec will change and the bastion will be re-created.

**Comment by EmilienM** [comment]
@lentzi90 when time permits, please review this one.

Thanks!

**Review by lentzi90 — COMMENTED** [review]
Looks good!
One question before I approve: What is the deal with the bastion patches in the e2e test kustomizations? Are we adding/enabling bastion there now? Was it disabled before?
Why are we enabling it now?

The reason I'm asking (besides just trying to understand it) is that I worry that the increased resource usage might cause issues in the CI.

**Comment by mdbooth** [comment]
> Looks good! One question before I approve: What is the deal with the bastion patches in the e2e test kustomizations? Are we adding/enabling bastion there now? Was it disabled before? Why are we enabling it now?
> 
> The reason I'm asking (besides just trying to understand it) is that I worry that the increased resource usage might cause issues in the CI.

Good catch! With a quick scan I can't currently see any e2e bastion tests, which feels like an obvious oversight. @EmilienM I'm guessing this is why you added them? Unfortunately I don't think this is safe to do for the reason @lentzi90 mentions. We should probably have a single, separate e2e bastion test.

Given the relatively limited scope of this PR I think I'd probably be ok opening a separate issue for that and just removing the bastion hosts in the e2e tests here. @lentzi90 what do you think?

**Comment by lentzi90** [comment]
> Given the relatively limited scope of this PR I think I'd probably be ok opening a separate issue for that and just removing the bastion hosts in the e2e tests here. @lentzi90 what do you think?

Sounds good to me! :+1:

**Comment by EmilienM** [comment]
This was a mistake on my side, I'll fix it. Nice catch and thanks for your review!

**Comment by EmilienM** [comment]
> Given the relatively limited scope of this PR I think I'd probably be ok opening a separate issue for that and just removing the bastion hosts in the e2e tests here.

Ok I'll do it.

**Comment by EmilienM** [comment]
I checked and the bastion is used to run `CollectMachineLog` as it goes through the bastion to reach the machines. But this function isn't called anywhere :shrug: ...

**Comment by EmilienM** [comment]
@lentzi90 @mdbooth I have a preference to address this via another PR, which I'm working on now.

**Comment by EmilienM** [comment]
@lentzi90 @mdbooth ready for review again. Thanks!

File: test/e2e/shared/common.go
Comment (lentzi90) [inline test/e2e/shared/common.go:218]: I'm confused about this. I mean the code makes sense to me but I'm confused about if we have bastions or not. Yesterday I thought that we didn't have bastions when I saw you added the bastion patches, but now when I see this and realize that we cannot collect logs without bastions, then we must have bastions currently because we definitely do have these logs!

What I see now in the tests run on this PR is that there are no longs for the machines (e.g. [here](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1796/pull-cluster-api-provider-openstack-e2e-test/1745523472215314432/artifacts/clusters/bootstrap/e2e-4p833k/machines/)). But in our periodic tests (e.g. [this](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-provider-openstack-e2e-test-main/1745503729274064896)) there [are logs](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/periodic-cluster-api-provider-openstack-e2e-test-main/1745503729274064896/artifacts/clusters/bootstrap/e2e-6urfd4/machines/) for some machines at least.
So now it seems to me that we do have bastions in at least some tests and this PR is perhaps removing them and consequently prevents log collection.

Comment (mdbooth) [reply test/e2e/shared/common.go:218]: I feel like I'm on the same rollercoaster. Weeeeeeeeeee.

**Review by lentzi90 — COMMENTED** [review]
/approve
The kustomize patches could be optimized a bit but I don't want to hold this back just because of that :slightly_smiling_face:

**Comment by EmilienM** [comment]
rebased

**Pull request merged**
